com.unity.perception/Runtime/Randomization/Randomizers/AssetSources/AssetSource.cs

```diff
@@ -0,0 +1,139 @@
+﻿using System;
+using UnityEngine.Perception.Randomization.Samplers;
+
+namespace UnityEngine.Perception.Randomization
+{
+    /// <summary>
+    /// AssetSources are used to load assets from a generically within a <see cref="Randomizers.Randomizer"/>
+    /// </summary>
+    /// <typeparam name="T">The type of asset to load</typeparam>
+    [Serializable]
+    public sealed class AssetSource<T> where T : Object
+    {
+        [SerializeReference] ArchetypeBase m_ArchetypeBase;
+
+        /// <summary>
+        /// The location to load assets from
+        /// </summary>
+        [SerializeReference] public AssetSourceLocation assetSourceLocation = new LocalAssetSourceLocation();
+
+        bool m_Initialized;
+        UniformSampler m_Sampler = new UniformSampler();
+
+        /// <summary>
+        /// The archetype used to preprocess assets from this source
+        /// </summary>
+        public Archetype<T> archetype
+        {
+            get => (Archetype<T>)m_ArchetypeBase;
+            set => m_ArchetypeBase = value;
+        }
+
+        /// <summary>
+        /// The number of assets available within this asset source
+        /// </summary>
+        public int Count => assetSourceLocation.Count;
+
+        /// <summary>
+        /// Execute setup steps for this AssetSource
+        /// </summary>
+        public void Initialize()
+        {
+            assetSourceLocation.Initialize(archetype);
+            m_Initialized = true;
+        }
+
+        /// <summary>
+        /// Returns the unprocessed asset loaded from the provided index
+        /// </summary>
+        /// <param name="index">The index of the asset to load</param>
+        /// <returns>The asset loaded at the provided index</returns>
+        public T LoadRawAsset(int index)
+        {
+            CheckIfInitialized();
+            return assetSourceLocation.LoadAsset<T>(index);
+        }
+
+        /// <summary>
+        /// Returns all unprocessed assets that can be loaded from this AssetSource
+        /// </summary>
+        /// <returns>All assets that can be loaded from this AssetSource</returns>
+        public T[] LoadAllRawAssets()
```

Comment:
It says "Returns all unprocessed assets" but it calls `archetype.Preprocess`. Intended?

Reply:
Oops, def not intended. Good catch.